legend-engine-core/legend-engine-core-language-pure/legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/SDLCLoader.java

```diff
@@ -148,30 +162,38 @@ private boolean isCacheableAlloySDLC(SDLC sdlc)
     @Override
     public PureModelContext cacheKey(PureModelContext context, Identity identity)
     {
-        if (isCacheablePureSDLC(((PureModelContextPointer) context).sdlcInfo))
+        if (context instanceof PureModelContextPointer)
         {
-            final Subject executionSubject = getSubject();
-            Function0<PureModelContext> pureModelContextFunction = () -> this.pureLoader.getCacheKey(context, identity, executionSubject);
-            return executionSubject == null ? pureModelContextFunction.value() : exec(executionSubject, pureModelContextFunction::value);
+            if (isCacheablePureSDLC(((PureModelContextPointer) context).sdlcInfo))
+            {
+                final Subject executionSubject = getSubject();
+                Function0<PureModelContext> pureModelContextFunction = () -> this.pureLoader.getCacheKey(context, identity, executionSubject);
+                return executionSubject == null ? pureModelContextFunction.value() : exec(executionSubject, pureModelContextFunction::value);
+            }
+            else
+            {
+                return this.alloyLoader.getCacheKey(context);
+            }
         }
-        else
+        else if (context instanceof PureModelContextCollection)
         {
-            return this.alloyLoader.getCacheKey(context);
+            PureModelContextCollection pureModelContextCollection = (PureModelContextCollection) context;
+            Assert.assertTrue(pureModelContextCollection.getContexts().stream().allMatch(pureModelContext -> pureModelContext instanceof PureModelContextPointer), () -> "Invalid type of PureModelContext in PureModelContextCollection for cacheKey");
```

Comment:
This constrains defeats one of the main purposes of the collection I wanted to introduce - improve studio flow. For example - studio can manage a collection, where an entry is a PMCD of the current project, and the other entries are pointer to the project dependencies.

Reply:
I will remove this constraint so it will help improve studio flow. I will also change the constraint added on shouldCache but since we cache compiled model only for PureModelContextPointer, we won't gain much there in case of studio.

Also, I have placed another check in SDLCFetcher [see here](https://github.com/finos/legend-engine/pull/2641/files#diff-5e7885c22761b0fcb1da9a3fb4d667c3c60def193a97f369f9e5aca01227af74R96). It's required as my implementation only uses alloy loaders and thus only AlloySDLC. If there's a use case to support sourcing from multiple loaders the logic would have to be changed accordingly.